packages/client-browser/src/connection/browser_connection.ts

```diff
@@ -0,0 +1,178 @@
+import type {
+  BaseQueryParams,
+  Connection,
+  ConnectionParams,
+  InsertParams,
+  InsertResult,
+  QueryResult,
+} from '@clickhouse/client-common/connection'
+import { getAsText, isStream } from '../utils'
+import {
+  getQueryId,
+  isSuccessfulResponse,
+  toSearchParams,
+  transformUrl,
+  withCompressionHeaders,
+  withHttpSettings,
+} from '@clickhouse/client-common/utils'
+import { parseError } from '@clickhouse/client-common/error'
+import type { URLSearchParams } from 'url'
+
+export class BrowserConnection implements Connection<ReadableStream> {
+  private readonly defaultHeaders: Record<string, string>
+  constructor(private readonly params: ConnectionParams) {
+    this.defaultHeaders = {
+      Authorization: `Basic ${btoa(`${params.username}:${params.password}`)}`,
+    }
+  }
+
+  async query(
+    params: BaseQueryParams
+  ): Promise<QueryResult<ReadableStream<Uint8Array>>> {
+    const query_id = getQueryId(params.query_id)
+    const clickhouse_settings = withHttpSettings(
+      params.clickhouse_settings,
+      this.params.compression.decompress_response
+    )
+    const searchParams = toSearchParams({
+      database: this.params.database,
+      clickhouse_settings,
+      query_params: params.query_params,
+      session_id: params.session_id,
+      query_id,
+    })
+    const response = await this.request({
+      body: params.query,
+      params,
+      searchParams,
+    })
+    return {
+      query_id,
+      stream: response.body || new ReadableStream<Uint8Array>(),
+    }
+  }
+
+  async exec(params: BaseQueryParams): Promise<QueryResult<ReadableStream>> {
+    const query_id = getQueryId(params.query_id)
+    const searchParams = toSearchParams({
+      database: this.params.database,
+      clickhouse_settings: params.clickhouse_settings,
+      query_params: params.query_params,
+      session_id: params.session_id,
+      query_id,
+    })
+    const response = await this.request({
+      body: params.query,
+      params,
+      searchParams,
+    })
+    return {
+      stream: response.body || new ReadableStream<Uint8Array>(),
+      query_id,
+    }
+  }
+
+  async insert<T = unknown>(
+    params: InsertParams<ReadableStream<T>>
+  ): Promise<InsertResult> {
+    const query_id = getQueryId(params.query_id)
+    const searchParams = toSearchParams({
+      database: this.params.database,
+      clickhouse_settings: params.clickhouse_settings,
+      query_params: params.query_params,
+      query: params.query,
+      session_id: params.session_id,
+      query_id,
+    })
+    await this.request({
+      body: params.values,
+      params,
+      searchParams,
+    })
+    return {
+      query_id,
+    }
+  }
+
+  async ping(): Promise<boolean> {
+    // TODO: catch an error and just log it, returning false?
```

Comment:
maybe it's not a bad idea, considering that request might fail due to CORS, for example

Reply:
Added a follow-up https://github.com/ClickHouse/clickhouse-js/issues/178